core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,200 @@
+      .set("spark.scheduler.executorTaskBlacklistTime", "1000")
+    sc = new SparkContext(conf)
+    val scheduler = new TaskSchedulerImpl(sc, 1)
+
+    val tracker = new BlacklistTracker(conf, clock)
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("1", "2"))
+    assert(tracker.executorBlacklist(scheduler, stage1, 2) === Set())
+
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("2"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(stage1, taskInfo_3_3_hostB, Success)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set("2"))
+
+    tracker.updateFailureExecutors(stage1, taskInfo_3_3_hostB, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 3) === Set("3"))
+    assert(tracker.nodeBlacklist() === Set())
+
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, Success)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+
+    // Task 2 on Stage 2 failed on Executor 2
+    tracker.updateFailureExecutors(stage2, taskInfo_2_2_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+    assert(tracker.executorBlacklist(scheduler, stage2, 1) === Set())
+    assert(tracker.executorBlacklist(scheduler, stage1, 2) === Set())
+    assert(tracker.executorBlacklist(scheduler, stage2, 2) === Set("2"))
+  }
+
+  test ("threshold strategy works") {
+    val conf = new SparkConf().setAppName("test").setMaster("local")
+      .set("spark.ui.enabled", "false")
+      .set("spark.scheduler.blacklist.strategy", "threshold")
+      .set("spark.scheduler.blacklist.threshold.maxFailureTaskNumber", "1")
+      .set("spark.scheduler.blacklist.threshold.maxBlackExecutorNumber", "1")
+
+    sc = new SparkContext(conf)
+    val scheduler = spy(new TaskSchedulerImpl(sc, 1))
+    when(scheduler.getExecutorsAliveOnHost("hostA")).thenReturn(Some(Set("1", "2", "4")))
+
+    val tracker = new BlacklistTracker(conf, clock)
+    // Task 1 failed on Executor 1
+    tracker.updateFailureExecutors(stage1, taskInfo_1_1_hostA, FAILURE)
+    assert(tracker.executorBlacklist(scheduler, stage1, 1) === Set())
+    // Task 2 failed on Executor 1
+    tracker.updateFailureExecutors(stage1, taskInfo_1_2_hostA, FAILURE)
+    // Executor 1 is now blacklisted from "all" tasks
+    (0 until 10).foreach { case i: Int =>
+      assert(tracker.executorBlacklist(scheduler, stage1, i) === Set("1"))
+    }
+    assert(tracker.nodeBlacklist() === Set())
+
+    // Task 1 and Task 2 failed on Executor 2
+    tracker.updateFailureExecutors(stage1, taskInfo_2_1_hostA, FAILURE)
+    tracker.updateFailureExecutors(stage1, taskInfo_2_2_hostA, FAILURE)
+    // Executor 2 is blacklisted and Executor 4 is blacklisted at the same time
```

Comment:
took me a second to understand this, can you change comment to:
Executor 2 is blacklisted, which leads to the node getting blacklisted so Executor 4 is blacklisted as well